File: MATTHEW/leetcode/AIcoding/3.DeepSeek-v3/erlang_solutions/2172-maximum-and-sum-of-array.erl
-module(solution).
-export([maximum_and_sum/2]).

maximum_and_sum(Nums, NumSlots) ->
    MaxMask = 1 bsl (2 * NumSlots),
    DP = array:new(MaxMask, {default, -1}),
    DP1 = array:set(0, 0, DP),
    solve(Nums, 0, DP1, NumSlots).

solve([], _, DP, _) ->
    lists:max(array:sparse_to_orddict(DP));
solve([Num | Nums], Mask, DP, NumSlots) ->
    NewDP = update_dp(Num, Mask, DP, NumSlots),
    solve(Nums, Mask, NewDP, NumSlots).

update_dp(Num, Mask, DP, NumSlots) ->
    lists:foldl(fun(Slot, AccDP) ->
        SlotMask = 3 bsl (2 * (Slot - 1)),
        Current = (Mask band SlotMask) bsr (2 * (Slot - 1)),
        if
            Current < 2 ->
                NewMask = Mask bor (1 bsl (2 * (Slot - 1))),
                CurrentVal = array:get(Mask, AccDP),
                NewVal = CurrentVal + (Num band Slot),
                case array:get(NewMask, AccDP) of
                    -1 -> array:set(NewMask, NewVal, AccDP);
                    ExistingVal when NewVal > ExistingVal -> array:set(NewMask, NewVal, AccDP);
                    _ -> AccDP
                end;
            true -> AccDP
        end
    end, DP, lists:seq(1, NumSlots)).

File: leetcode/AIcoding/4.GPT-4o/go_solutions/2172-maximum-and-sum-of-array.go
func maximumANDSum(nums []int, numSlots int) int {
    n := len(nums)
    maxMask := 1 << (numSlots * 2)
    dp := make([]int, maxMask)

    for mask := 0; mask < maxMask; mask++ {
        count := make([]int, numSlots)
        for i := 0; i < numSlots; i++ {
            count[i] = (mask >> (i * 2)) & 3
        }
        for i := 0; i < n; i++ {
            for j := 0; j < numSlots; j++ {
                if count[j] < 2 {
                    newMask := mask | (1 << (j * 2))
                    dp[newMask] = max(dp[newMask], dp[mask]+(nums[i]&^(1<<j)))
                }
            }
        }
    }
    return dp[maxMask-1]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}